# ETL

Carregando os dados

In [11]:
import numpy as np
import pandas as pd
import plotly.express as px
url = 'https://raw.githubusercontent.com/alura-cursos/challenge2-data-science/refs/heads/main/TelecomX_Data.json'
dados = pd.read_json(url)
dados.head()

customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4  {'Contract': 'Month-to-month', 'PaperlessBilli...

Visualizando as colunas e seus tipos

In [12]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   customerID  7267 non-null   object
 1   Churn       7267 non-null   object
 2   customer    7267 non-null   object
 3   phone       7267 non-null   object
 4   internet    7267 non-null   object
 5   account     7267 non-null   object
dtypes: object(6)
memory usage: 340.8+ KB


In [13]:
list(dados.columns)

['customerID', 'Churn', 'customer', 'phone', 'internet', 'account']

Desaninhando o JSON

In [14]:
# Lista das colunas com dados aninhados
colunas_aninhadas = ['customer', 'phone', 'internet', 'account']
colunas_desaninhadas = ['customerID', 'Churn']
dados_desaninhados = pd.DataFrame()

for coluna in colunas_desaninhadas:
    dados_desaninhados[coluna] = dados[coluna]

# iteração sob a lista de colunas para normalizar os dados, com concatenação de cada nova coluna ao df com eixo = 1 para colunas
for coluna in colunas_aninhadas:
    normalizado = pd.json_normalize(dados[coluna])
    dados_desaninhados = pd.concat([dados_desaninhados, normalizado], axis=1)


dados_desaninhados



customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0     0002-ORFBO    No  Female              0     Yes        Yes       9   
1     0003-MKNFE    No    Male              0      No         No       9   
2     0004-TLHLJ   Yes    Male              0      No         No       4   
3     0011-IGKFF   Yes    Male              1     Yes         No      13   
4     0013-EXCHZ   Yes  Female              1     Yes         No       3   
...          ...   ...     ...            ...     ...        ...     ...   
7262  9987-LUTYD    No  Female              0      No         No      13   
7263  9992-RRAMN   Yes    Male              0     Yes         No      22   
7264  9992-UJOEL    No    Male              0      No         No       2   
7265  9993-LHIEB    No    Male              0     Yes        Yes      67   
7266  9995-HOTOH    No    Male              0     Yes        Yes      63   

     PhoneService     MultipleLines InternetService  ... OnlineBackup  \
0             Yes                No             DSL  ...          Yes   
1             Yes               Yes             DSL  ...           No   
2             Yes                No     Fiber optic  ...           No   
3             Yes                No     Fiber optic  ...          Yes   
4             Yes                No     Fiber optic  ...           No   
...           ...               ...             ...  ...          ...   
7262          Yes                No             DSL  ...           No   
7263          Yes               Yes     Fiber optic  ...           No   
7264          Yes                No             DSL  ...          Yes   
7265          Yes                No             DSL  ...           No   
7266           No  No phone service             DSL  ...          Yes   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                  No         Yes         Yes              No        One year   
1                  No          No          No             Yes  Month-to-month   
2                 Yes          No          No              No  Month-to-month   
3                 Yes          No         Yes             Yes  Month-to-month   
4                  No         Yes         Yes              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7262               No         Yes          No              No        One year   
7263               No          No          No             Yes  Month-to-month   
7264               No          No          No              No  Month-to-month   
7265              Yes         Yes          No             Yes        Two year   
7266              Yes          No         Yes             Yes        Two year   

     PaperlessBilling     PaymentMethod Charges.Monthly  Charges.Total  
0                 Yes      Mailed check           65.60          593.3  
1                  No      Mailed check           59.90          542.4  
2                 Yes  Electronic check           73.90         280.85  
3                 Yes  Electronic check           98.00        1237.85  
4                 Yes      Mailed check           83.90          267.4  
...               ...               ...             ...            ...  
7262               No      Mailed check           55.15          742.9  
7263              Yes  Electronic check           85.10         1873.7  
7264              Yes      Mailed check           50.30          92.75  
7265               No      Mailed check           67.85        4627.65  
7266               No  Electronic check           59.00         3707.6  

[7267 rows x 21 columns]

Verificando os tipos das novas colunas

In [15]:
dados_desaninhados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7267 non-null   object 
 1   Churn             7267 non-null   object 
 2   gender            7267 non-null   object 
 3   SeniorCitizen     7267 non-null   int64  
 4   Partner           7267 non-null   object 
 5   Dependents        7267 non-null   object 
 6   tenure            7267 non-null   int64  
 7   PhoneService      7267 non-null   object 
 8   MultipleLines     7267 non-null   object 
 9   InternetService   7267 non-null   object 
 10  OnlineSecurity    7267 non-null   object 
 11  OnlineBackup      7267 non-null   object 
 12  DeviceProtection  7267 non-null   object 
 13  TechSupport       7267 non-null   object 
 14  StreamingTV       7267 non-null   object 
 15  StreamingMovies   7267 non-null   object 
 16  Contract          7267 non-null   object 


Transformando SenionCitizen de int64 para bool

In [16]:
dados_desaninhados['SeniorCitizen'] = dados_desaninhados['SeniorCitizen'].astype(bool)
dados_desaninhados

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0     0002-ORFBO    No  Female          False     Yes        Yes       9   
1     0003-MKNFE    No    Male          False      No         No       9   
2     0004-TLHLJ   Yes    Male          False      No         No       4   
3     0011-IGKFF   Yes    Male           True     Yes         No      13   
4     0013-EXCHZ   Yes  Female           True     Yes         No       3   
...          ...   ...     ...            ...     ...        ...     ...   
7262  9987-LUTYD    No  Female          False      No         No      13   
7263  9992-RRAMN   Yes    Male          False     Yes         No      22   
7264  9992-UJOEL    No    Male          False      No         No       2   
7265  9993-LHIEB    No    Male          False     Yes        Yes      67   
7266  9995-HOTOH    No    Male          False     Yes        Yes      63   

     PhoneService     MultipleLines InternetService  ... OnlineBackup  \
0             Yes                No             DSL  ...          Yes   
1             Yes               Yes             DSL  ...           No   
2             Yes                No     Fiber optic  ...           No   
3             Yes                No     Fiber optic  ...          Yes   
4             Yes                No     Fiber optic  ...           No   
...           ...               ...             ...  ...          ...   
7262          Yes                No             DSL  ...           No   
7263          Yes               Yes     Fiber optic  ...           No   
7264          Yes                No             DSL  ...          Yes   
7265          Yes                No             DSL  ...           No   
7266           No  No phone service             DSL  ...          Yes   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                  No         Yes         Yes              No        One year   
1                  No          No          No             Yes  Month-to-month   
2                 Yes          No          No              No  Month-to-month   
3                 Yes          No         Yes             Yes  Month-to-month   
4                  No         Yes         Yes              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7262               No         Yes          No              No        One year   
7263               No          No          No             Yes  Month-to-month   
7264               No          No          No              No  Month-to-month   
7265              Yes         Yes          No             Yes        Two year   
7266              Yes          No         Yes             Yes        Two year   

     PaperlessBilling     PaymentMethod Charges.Monthly  Charges.Total  
0                 Yes      Mailed check           65.60          593.3  
1                  No      Mailed check           59.90          542.4  
2                 Yes  Electronic check           73.90         280.85  
3                 Yes  Electronic check           98.00        1237.85  
4                 Yes      Mailed check           83.90          267.4  
...               ...               ...             ...            ...  
7262               No      Mailed check           55.15          742.9  
7263              Yes  Electronic check           85.10         1873.7  
7264              Yes      Mailed check           50.30          92.75  
7265               No      Mailed check           67.85        4627.65  
7266               No  Electronic check           59.00         3707.6  

[7267 rows x 21 columns]

Alterando nome de algumas colunas para facilitar querys. Eu já havia alterado algumas no decorrer do notebook, mas voltei e decidi alterar todas colunas aqui antes de começar a atratar os dados_desaninhados para ficar mais organizado

In [22]:
dados_desaninhados.rename(columns={'seniorcitizen': 'is_senior_citizen'}, inplace=True)
dados_desaninhados.rename(columns={'phoneservice': 'phone_service'}, inplace=True)
dados_desaninhados.rename(columns={'multiplelines': 'multiple_lines'}, inplace=True)
dados_desaninhados.rename(columns={'internetservice': 'internet_service'}, inplace=True)
dados_desaninhados.rename(columns={'onlinesecurity': 'online_security'}, inplace=True)
dados_desaninhados.rename(columns={'onlinebackup': 'online_backup'}, inplace=True)
dados_desaninhados.rename(columns={'deviceprotection': 'device_protection'}, inplace=True)
dados_desaninhados.rename(columns={'techsupport': 'tech_support'}, inplace=True)
dados_desaninhados.rename(columns={'streamingtv': 'streaming_tv'}, inplace=True)
dados_desaninhados.rename(columns={'streamingmovies': 'streaming_movies'}, inplace=True)
dados_desaninhados.rename(columns={'paperlessbilling': 'paperless_billing'}, inplace=True)
dados_desaninhados.rename(columns={'paymentmethod': 'payment_method'}, inplace=True)
dados_desaninhados.rename(columns={'charges.monthly': 'monthly_charges'}, inplace=True)
dados_desaninhados.rename(columns={'charges.total': 'total_charges'}, inplace=True)
dados_desaninhados.rename(columns={'customerid': 'customer_id'}, inplace=True)

dados_desaninhados.head()

customer_id churn  gender  is_senior_citizen partner dependents  tenure  \
0  0002-ORFBO    No  Female              False     Yes        Yes       9   
1  0003-MKNFE    No    Male              False      No         No       9   
2  0004-TLHLJ   Yes    Male              False      No         No       4   
3  0011-IGKFF   Yes    Male               True     Yes         No      13   
4  0013-EXCHZ   Yes  Female               True     Yes         No       3   

  phone_service multiple_lines internet_service  ... online_backup  \
0           Yes             No              DSL  ...           Yes   
1           Yes            Yes              DSL  ...            No   
2           Yes             No      Fiber optic  ...            No   
3           Yes             No      Fiber optic  ...           Yes   
4           Yes             No      Fiber optic  ...            No   

  device_protection tech_support streaming_tv streaming_movies  \
0                No          Yes          Yes               No   
1                No           No           No              Yes   
2               Yes           No           No               No   
3               Yes           No          Yes              Yes   
4                No          Yes          Yes               No   

         contract paperless_billing    payment_method monthly_charges  \
0        One year               Yes      Mailed check            65.6   
1  Month-to-month                No      Mailed check            59.9   
2  Month-to-month               Yes  Electronic check            73.9   
3  Month-to-month               Yes  Electronic check            98.0   
4  Month-to-month               Yes      Mailed check            83.9   

   total_charges  
0          593.3  
1          542.4  
2         280.85  
3        1237.85  
4          267.4  

[5 rows x 21 columns]

Padronizando nome das colunas

In [23]:
dados_desaninhados.columns = dados_desaninhados.columns.str.lower()
dados_desaninhados.head()

customer_id churn  gender  is_senior_citizen partner dependents  tenure  \
0  0002-ORFBO    No  Female              False     Yes        Yes       9   
1  0003-MKNFE    No    Male              False      No         No       9   
2  0004-TLHLJ   Yes    Male              False      No         No       4   
3  0011-IGKFF   Yes    Male               True     Yes         No      13   
4  0013-EXCHZ   Yes  Female               True     Yes         No       3   

  phone_service multiple_lines internet_service  ... online_backup  \
0           Yes             No              DSL  ...           Yes   
1           Yes            Yes              DSL  ...            No   
2           Yes             No      Fiber optic  ...            No   
3           Yes             No      Fiber optic  ...           Yes   
4           Yes             No      Fiber optic  ...            No   

  device_protection tech_support streaming_tv streaming_movies  \
0                No          Yes          Yes               No   
1                No           No           No              Yes   
2               Yes           No           No               No   
3               Yes           No          Yes              Yes   
4                No          Yes          Yes               No   

         contract paperless_billing    payment_method monthly_charges  \
0        One year               Yes      Mailed check            65.6   
1  Month-to-month                No      Mailed check            59.9   
2  Month-to-month               Yes  Electronic check            73.9   
3  Month-to-month               Yes  Electronic check            98.0   
4  Month-to-month               Yes      Mailed check            83.9   

   total_charges  
0          593.3  
1          542.4  
2         280.85  
3        1237.85  
4          267.4  

[5 rows x 21 columns]

Convertendo TotalCharges para float64 (Erro pois existem campos com valores incoerentes)
Campos retirados e conversão feita

In [29]:
dados_desaninhados['total_charges'] = dados_desaninhados['total_charges'].astype(float)
dados_desaninhados.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7256 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   customer_id        7256 non-null   object 
 1   churn              7256 non-null   object 
 2   gender             7256 non-null   object 
 3   is_senior_citizen  7256 non-null   bool   
 4   partner            7256 non-null   object 
 5   dependents         7256 non-null   object 
 6   tenure             7256 non-null   int64  
 7   phone_service      7256 non-null   object 
 8   multiple_lines     7256 non-null   object 
 9   internet_service   7256 non-null   object 
 10  online_security    7256 non-null   object 
 11  online_backup      7256 non-null   object 
 12  device_protection  7256 non-null   object 
 13  tech_support       7256 non-null   object 
 14  streaming_tv       7256 non-null   object 
 15  streaming_movies   7256 non-null   object 
 16  contract           7256 non-n

Identificando e quantificando quais registros possuem inconsistencias

In [26]:
dados_desaninhados[dados_desaninhados['total_charges'] == ' '].index

Index([975, 1775, 1955, 2075, 2232, 2308, 2930, 3134, 3203, 4169, 5599], dtype='int64')

Retirando dados (menos 11 registros)

In [27]:
dados_desaninhados = dados_desaninhados.drop(dados_desaninhados[dados_desaninhados['total_charges'] == ' '].index)
dados_desaninhados.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7256 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   customer_id        7256 non-null   object 
 1   churn              7256 non-null   object 
 2   gender             7256 non-null   object 
 3   is_senior_citizen  7256 non-null   bool   
 4   partner            7256 non-null   object 
 5   dependents         7256 non-null   object 
 6   tenure             7256 non-null   int64  
 7   phone_service      7256 non-null   object 
 8   multiple_lines     7256 non-null   object 
 9   internet_service   7256 non-null   object 
 10  online_security    7256 non-null   object 
 11  online_backup      7256 non-null   object 
 12  device_protection  7256 non-null   object 
 13  tech_support       7256 non-null   object 
 14  streaming_tv       7256 non-null   object 
 15  streaming_movies   7256 non-null   object 
 16  contract           7256 non-n

In [28]:
dados_desaninhados.head()

customer_id churn  gender  is_senior_citizen partner dependents  tenure  \
0  0002-ORFBO    No  Female              False     Yes        Yes       9   
1  0003-MKNFE    No    Male              False      No         No       9   
2  0004-TLHLJ   Yes    Male              False      No         No       4   
3  0011-IGKFF   Yes    Male               True     Yes         No      13   
4  0013-EXCHZ   Yes  Female               True     Yes         No       3   

  phone_service multiple_lines internet_service  ... online_backup  \
0           Yes             No              DSL  ...           Yes   
1           Yes            Yes              DSL  ...            No   
2           Yes             No      Fiber optic  ...            No   
3           Yes             No      Fiber optic  ...           Yes   
4           Yes             No      Fiber optic  ...            No   

  device_protection tech_support streaming_tv streaming_movies  \
0                No          Yes          Yes               No   
1                No           No           No              Yes   
2               Yes           No           No               No   
3               Yes           No          Yes              Yes   
4                No          Yes          Yes               No   

         contract paperless_billing    payment_method monthly_charges  \
0        One year               Yes      Mailed check            65.6   
1  Month-to-month                No      Mailed check            59.9   
2  Month-to-month               Yes  Electronic check            73.9   
3  Month-to-month               Yes  Electronic check            98.0   
4  Month-to-month               Yes      Mailed check            83.9   

   total_charges  
0          593.3  
1          542.4  
2         280.85  
3        1237.85  
4          267.4  

[5 rows x 21 columns]

# Procurando inconsistencias

Dados ausentes(NA) não existem

In [30]:

for coluna in dados_desaninhados.columns:
    print(f'{coluna}: {dados_desaninhados[coluna].isna().sum()}')

customer_id: 0
churn: 0
gender: 0
is_senior_citizen: 0
partner: 0
dependents: 0
tenure: 0
phone_service: 0
multiple_lines: 0
internet_service: 0
online_security: 0
online_backup: 0
device_protection: 0
tech_support: 0
streaming_tv: 0
streaming_movies: 0
contract: 0
paperless_billing: 0
payment_method: 0
monthly_charges: 0
total_charges: 0


Dados vazios ('') existem na coluna Churn, com 224 ocorrencias. Os dados podem ser excluidos ou preenchidos aleatoriamente com base na média/moda. Escolhi excluir os dados por representarem apenas 3% da amostra. Dessa forma, os registros mantidos são confiáveis e fidedignos.

In [31]:
for coluna in dados_desaninhados.columns:
  indice_coluna = dados_desaninhados[dados_desaninhados[coluna] == ''].index
  print(f' coluna {coluna}: indice{indice_coluna}')

 coluna customer_id: indiceIndex([], dtype='int64')
 coluna churn: indiceIndex([  30,   75,   96,   98,  175,  219,  312,  351,  368,  374,
       ...
       7058, 7061, 7062, 7143, 7155, 7158, 7180, 7211, 7239, 7247],
      dtype='int64', length=224)
 coluna gender: indiceIndex([], dtype='int64')
 coluna is_senior_citizen: indiceIndex([], dtype='int64')
 coluna partner: indiceIndex([], dtype='int64')
 coluna dependents: indiceIndex([], dtype='int64')
 coluna tenure: indiceIndex([], dtype='int64')
 coluna phone_service: indiceIndex([], dtype='int64')
 coluna multiple_lines: indiceIndex([], dtype='int64')
 coluna internet_service: indiceIndex([], dtype='int64')
 coluna online_security: indiceIndex([], dtype='int64')
 coluna online_backup: indiceIndex([], dtype='int64')
 coluna device_protection: indiceIndex([], dtype='int64')
 coluna tech_support: indiceIndex([], dtype='int64')
 coluna streaming_tv: indiceIndex([], dtype='int64')
 coluna streaming_movies: indiceIndex([], dtype='int64')


In [33]:
dados_desaninhados = dados_desaninhados.drop(dados_desaninhados[dados_desaninhados['churn'] == ''].index)
dados_desaninhados.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7032 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   customer_id        7032 non-null   object 
 1   churn              7032 non-null   object 
 2   gender             7032 non-null   object 
 3   is_senior_citizen  7032 non-null   bool   
 4   partner            7032 non-null   object 
 5   dependents         7032 non-null   object 
 6   tenure             7032 non-null   int64  
 7   phone_service      7032 non-null   object 
 8   multiple_lines     7032 non-null   object 
 9   internet_service   7032 non-null   object 
 10  online_security    7032 non-null   object 
 11  online_backup      7032 non-null   object 
 12  device_protection  7032 non-null   object 
 13  tech_support       7032 non-null   object 
 14  streaming_tv       7032 non-null   object 
 15  streaming_movies   7032 non-null   object 
 16  contract           7032 non-n

In [35]:
dados_desaninhados.head()

customer_id churn  gender  is_senior_citizen partner dependents  tenure  \
0  0002-ORFBO    No  Female              False     Yes        Yes       9   
1  0003-MKNFE    No    Male              False      No         No       9   
2  0004-TLHLJ   Yes    Male              False      No         No       4   
3  0011-IGKFF   Yes    Male               True     Yes         No      13   
4  0013-EXCHZ   Yes  Female               True     Yes         No       3   

  phone_service multiple_lines internet_service  ... online_backup  \
0           Yes             No              DSL  ...           Yes   
1           Yes            Yes              DSL  ...            No   
2           Yes             No      Fiber optic  ...            No   
3           Yes             No      Fiber optic  ...           Yes   
4           Yes             No      Fiber optic  ...            No   

  device_protection tech_support streaming_tv streaming_movies  \
0                No          Yes          Yes               No   
1                No           No           No              Yes   
2               Yes           No           No               No   
3               Yes           No          Yes              Yes   
4                No          Yes          Yes               No   

         contract paperless_billing    payment_method monthly_charges  \
0        One year               Yes      Mailed check            65.6   
1  Month-to-month                No      Mailed check            59.9   
2  Month-to-month               Yes  Electronic check            73.9   
3  Month-to-month               Yes  Electronic check            98.0   
4  Month-to-month               Yes      Mailed check            83.9   

   total_charges  
0         593.30  
1         542.40  
2         280.85  
3        1237.85  
4         267.40  

[5 rows x 21 columns]

Verificando os registros existentes em cada coluna em busca de inconsistencias

In [36]:
for coluna in dados_desaninhados.columns:
  if coluna != 'customerID':
    print(dados_desaninhados[coluna].value_counts())

customer_id
9995-HOTOH    1
0002-ORFBO    1
0003-MKNFE    1
0004-TLHLJ    1
0011-IGKFF    1
             ..
0018-NYROU    1
0017-IUDMW    1
0017-DINOC    1
0016-QLJIS    1
0015-UOCOJ    1
Name: count, Length: 7032, dtype: int64
churn
No     5163
Yes    1869
Name: count, dtype: int64
gender
Male      3549
Female    3483
Name: count, dtype: int64
is_senior_citizen
False    5890
True     1142
Name: count, dtype: int64
partner
No     3639
Yes    3393
Name: count, dtype: int64
dependents
No     4933
Yes    2099
Name: count, dtype: int64
tenure
1     613
72    362
2     238
3     200
4     176
     ... 
38     59
28     57
39     56
44     51
36     50
Name: count, Length: 72, dtype: int64
phone_service
Yes    6352
No      680
Name: count, dtype: int64
multiple_lines
No                  3385
Yes                 2967
No phone service     680
Name: count, dtype: int64
internet_service
Fiber optic    3096
DSL            2416
No             1520
Name: count, dtype: int64
online_security
No      

Querys

In [37]:
resultado_query = dados_desaninhados.query('tenure == 1')
print(resultado_query[['customer_id', 'tenure', 'monthly_charges', 'total_charges']])

     customer_id  tenure  monthly_charges  total_charges
17    0021-IKXGC       1            72.10          72.10
19    0023-HGHWL       1            25.10          25.10
25    0032-PGELS       1            30.50          30.50
49    0082-LDZUE       1            44.30          44.30
64    0107-WESLM       1            19.85          19.85
...          ...     ...              ...            ...
7201  9907-SWKKF       1            25.05          25.05
7229  9940-RHLFB       1            75.30          75.30
7244  9962-BFPDU       1            20.05          20.05
7257  9975-SKRNR       1            18.90          18.90
7260  9985-MWVIX       1            70.15          70.15

[613 rows x 4 columns]


In [38]:
dados_desaninhados.is_senior_citizen.value_counts()

is_senior_citizen
False    5890
True     1142
Name: count, dtype: int64

Padronizando is_senior_citizen para padrao da tabela(Yes, No)

In [40]:
dados_desaninhados['is_senior_citizen'] = dados_desaninhados['is_senior_citizen'].map({True: 'Yes', False: 'No'})
dados_desaninhados.head()

customer_id churn  gender is_senior_citizen partner dependents  tenure  \
0  0002-ORFBO    No  Female                No     Yes        Yes       9   
1  0003-MKNFE    No    Male                No      No         No       9   
2  0004-TLHLJ   Yes    Male                No      No         No       4   
3  0011-IGKFF   Yes    Male               Yes     Yes         No      13   
4  0013-EXCHZ   Yes  Female               Yes     Yes         No       3   

  phone_service multiple_lines internet_service  ... online_backup  \
0           Yes             No              DSL  ...           Yes   
1           Yes            Yes              DSL  ...            No   
2           Yes             No      Fiber optic  ...            No   
3           Yes             No      Fiber optic  ...           Yes   
4           Yes             No      Fiber optic  ...            No   

  device_protection tech_support streaming_tv streaming_movies  \
0                No          Yes          Yes               No   
1                No           No           No              Yes   
2               Yes           No           No               No   
3               Yes           No          Yes              Yes   
4                No          Yes          Yes               No   

         contract paperless_billing    payment_method monthly_charges  \
0        One year               Yes      Mailed check            65.6   
1  Month-to-month                No      Mailed check            59.9   
2  Month-to-month               Yes  Electronic check            73.9   
3  Month-to-month               Yes  Electronic check            98.0   
4  Month-to-month               Yes      Mailed check            83.9   

   total_charges  
0         593.30  
1         542.40  
2         280.85  
3        1237.85  
4         267.40  

[5 rows x 21 columns]

# Coluna de Contas diárias

Criando novo DF dados_tratados

In [41]:
dados_tratados = dados_desaninhados.copy()
dados_tratados.head()

customer_id churn  gender is_senior_citizen partner dependents  tenure  \
0  0002-ORFBO    No  Female                No     Yes        Yes       9   
1  0003-MKNFE    No    Male                No      No         No       9   
2  0004-TLHLJ   Yes    Male                No      No         No       4   
3  0011-IGKFF   Yes    Male               Yes     Yes         No      13   
4  0013-EXCHZ   Yes  Female               Yes     Yes         No       3   

  phone_service multiple_lines internet_service  ... online_backup  \
0           Yes             No              DSL  ...           Yes   
1           Yes            Yes              DSL  ...            No   
2           Yes             No      Fiber optic  ...            No   
3           Yes             No      Fiber optic  ...           Yes   
4           Yes             No      Fiber optic  ...            No   

  device_protection tech_support streaming_tv streaming_movies  \
0                No          Yes          Yes               No   
1                No           No           No              Yes   
2               Yes           No           No               No   
3               Yes           No          Yes              Yes   
4                No          Yes          Yes               No   

         contract paperless_billing    payment_method monthly_charges  \
0        One year               Yes      Mailed check            65.6   
1  Month-to-month                No      Mailed check            59.9   
2  Month-to-month               Yes  Electronic check            73.9   
3  Month-to-month               Yes  Electronic check            98.0   
4  Month-to-month               Yes      Mailed check            83.9   

   total_charges  
0         593.30  
1         542.40  
2         280.85  
3        1237.85  
4         267.40  

[5 rows x 21 columns]

Criando nova coluna e assumindo os meses com 30 dias para facilitar a conta

In [42]:
dados_tratados['daily_charges'] = dados_tratados['monthly_charges'] / 30.0
dados_tratados.daily_charges

0       2.186667
1       1.996667
2       2.463333
3       3.266667
4       2.796667
          ...   
7262    1.838333
7263    2.836667
7264    1.676667
7265    2.261667
7266    1.966667
Name: daily_charges, Length: 7032, dtype: float64

arredondando as casas para padronização

In [43]:
dados_tratados['daily_charges'] = dados_tratados['daily_charges'].round(2)
dados_tratados.head()

customer_id churn  gender is_senior_citizen partner dependents  tenure  \
0  0002-ORFBO    No  Female                No     Yes        Yes       9   
1  0003-MKNFE    No    Male                No      No         No       9   
2  0004-TLHLJ   Yes    Male                No      No         No       4   
3  0011-IGKFF   Yes    Male               Yes     Yes         No      13   
4  0013-EXCHZ   Yes  Female               Yes     Yes         No       3   

  phone_service multiple_lines internet_service  ... device_protection  \
0           Yes             No              DSL  ...                No   
1           Yes            Yes              DSL  ...                No   
2           Yes             No      Fiber optic  ...               Yes   
3           Yes             No      Fiber optic  ...               Yes   
4           Yes             No      Fiber optic  ...                No   

  tech_support streaming_tv streaming_movies        contract  \
0          Yes          Yes               No        One year   
1           No           No              Yes  Month-to-month   
2           No           No               No  Month-to-month   
3           No          Yes              Yes  Month-to-month   
4          Yes          Yes               No  Month-to-month   

  paperless_billing    payment_method monthly_charges total_charges  \
0               Yes      Mailed check            65.6        593.30   
1                No      Mailed check            59.9        542.40   
2               Yes  Electronic check            73.9        280.85   
3               Yes  Electronic check            98.0       1237.85   
4               Yes      Mailed check            83.9        267.40   

   daily_charges  
0           2.19  
1           2.00  
2           2.46  
3           3.27  
4           2.80  

[5 rows x 22 columns]

Coluna daily_charges criada

In [45]:
dados_tratados.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7032 entries, 0 to 7266
Data columns (total 22 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   customer_id        7032 non-null   object 
 1   churn              7032 non-null   object 
 2   gender             7032 non-null   object 
 3   is_senior_citizen  7032 non-null   object 
 4   partner            7032 non-null   object 
 5   dependents         7032 non-null   object 
 6   tenure             7032 non-null   int64  
 7   phone_service      7032 non-null   object 
 8   multiple_lines     7032 non-null   object 
 9   internet_service   7032 non-null   object 
 10  online_security    7032 non-null   object 
 11  online_backup      7032 non-null   object 
 12  device_protection  7032 non-null   object 
 13  tech_support       7032 non-null   object 
 14  streaming_tv       7032 non-null   object 
 15  streaming_movies   7032 non-null   object 
 16  contract           7032 non-n

Salvando DF para continuar depois

In [46]:
dados_tratados.to_csv('dados_tratados.csv', index=False)

# Padronização e Transformação de Dados

Novo DF dados_padronizados

In [79]:
url = 'https://raw.githubusercontent.com/mmnasciment/TelecomX_CustomersChurn/refs/heads/main/dados_tratados.csv'
dados_padronizados = pd.read_csv(url)
dados_padronizados

customer_id churn  gender is_senior_citizen partner dependents  tenure  \
0     0002-ORFBO    No  Female                No     Yes        Yes       9   
1     0003-MKNFE    No    Male                No      No         No       9   
2     0004-TLHLJ   Yes    Male                No      No         No       4   
3     0011-IGKFF   Yes    Male               Yes     Yes         No      13   
4     0013-EXCHZ   Yes  Female               Yes     Yes         No       3   
...          ...   ...     ...               ...     ...        ...     ...   
7027  9987-LUTYD    No  Female                No      No         No      13   
7028  9992-RRAMN   Yes    Male                No     Yes         No      22   
7029  9992-UJOEL    No    Male                No      No         No       2   
7030  9993-LHIEB    No    Male                No     Yes        Yes      67   
7031  9995-HOTOH    No    Male                No     Yes        Yes      63   

     phone_service    multiple_lines internet_service  ... device_protection  \
0              Yes                No              DSL  ...                No   
1              Yes               Yes              DSL  ...                No   
2              Yes                No      Fiber optic  ...               Yes   
3              Yes                No      Fiber optic  ...               Yes   
4              Yes                No      Fiber optic  ...                No   
...            ...               ...              ...  ...               ...   
7027           Yes                No              DSL  ...                No   
7028           Yes               Yes      Fiber optic  ...                No   
7029           Yes                No              DSL  ...                No   
7030           Yes                No              DSL  ...               Yes   
7031            No  No phone service              DSL  ...               Yes   

     tech_support streaming_tv streaming_movies        contract  \
0             Yes          Yes               No        One year   
1              No           No              Yes  Month-to-month   
2              No           No               No  Month-to-month   
3              No          Yes              Yes  Month-to-month   
4             Yes          Yes               No  Month-to-month   
...           ...          ...              ...             ...   
7027          Yes           No               No        One year   
7028           No           No              Yes  Month-to-month   
7029           No           No               No  Month-to-month   
7030          Yes           No              Yes        Two year   
7031           No          Yes              Yes        Two year   

     paperless_billing    payment_method monthly_charges total_charges  \
0                  Yes      Mailed check           65.60        593.30   
1                   No      Mailed check           59.90        542.40   
2                  Yes  Electronic check           73.90        280.85   
3                  Yes  Electronic check           98.00       1237.85   
4                  Yes      Mailed check           83.90        267.40   
...                ...               ...             ...           ...   
7027                No      Mailed check           55.15        742.90   
7028               Yes  Electronic check           85.10       1873.70   
7029               Yes      Mailed check           50.30         92.75   
7030                No      Mailed check           67.85       4627.65   
7031                No  Electronic check           59.00       3707.60   

      daily_charges  
0              2.19  
1              2.00  
2              2.46  
3              3.27  
4              2.80  
...             ...  
7027           1.84  
7028           2.84  
7029           1.68  
7030           2.26  
7031           1.97  

[7032 rows x 22 columns]

Padronização dos registros (garantir que todas colunas com "resposta" binária sigam o padrão binário Yes=1, No=0)

In [80]:
dados_padronizados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7032 entries, 0 to 7031
Data columns (total 22 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   customer_id        7032 non-null   object 
 1   churn              7032 non-null   object 
 2   gender             7032 non-null   object 
 3   is_senior_citizen  7032 non-null   object 
 4   partner            7032 non-null   object 
 5   dependents         7032 non-null   object 
 6   tenure             7032 non-null   int64  
 7   phone_service      7032 non-null   object 
 8   multiple_lines     7032 non-null   object 
 9   internet_service   7032 non-null   object 
 10  online_security    7032 non-null   object 
 11  online_backup      7032 non-null   object 
 12  device_protection  7032 non-null   object 
 13  tech_support       7032 non-null   object 
 14  streaming_tv       7032 non-null   object 
 15  streaming_movies   7032 non-null   object 
 16  contract           7032 

iterando sobre os valores de colunas tipo object

In [81]:
for coluna in dados_padronizados.columns:
  if coluna != 'customer_id' and coluna != 'tenure' and coluna != 'monthly_charges' and coluna != 'total_charges' and coluna != 'daily_charges':
    print(dados_padronizados[coluna].value_counts())

churn
No     5163
Yes    1869
Name: count, dtype: int64
gender
Male      3549
Female    3483
Name: count, dtype: int64
is_senior_citizen
No     5890
Yes    1142
Name: count, dtype: int64
partner
No     3639
Yes    3393
Name: count, dtype: int64
dependents
No     4933
Yes    2099
Name: count, dtype: int64
phone_service
Yes    6352
No      680
Name: count, dtype: int64
multiple_lines
No                  3385
Yes                 2967
No phone service     680
Name: count, dtype: int64
internet_service
Fiber optic    3096
DSL            2416
No             1520
Name: count, dtype: int64
online_security
No                     3497
Yes                    2015
No internet service    1520
Name: count, dtype: int64
online_backup
No                     3087
Yes                    2425
No internet service    1520
Name: count, dtype: int64
device_protection
No                     3094
Yes                    2418
No internet service    1520
Name: count, dtype: int64
tech_support
No                  

Será utilizado o metodo 1-de-c para padronizar as colunas para criar um arquivo com formato binário

In [82]:
lista_col_binaria = []
lista_col_nao_binaria = []

for coluna in dados_padronizados.columns:
  if coluna == 'customer_id':
    continue

  qntd_valores_unicos = dados_padronizados[coluna].nunique()
  if qntd_valores_unicos == 2:
    lista_col_binaria.append(coluna)
  else:
    lista_col_nao_binaria.append(coluna)

print(lista_col_binaria)
print(lista_col_nao_binaria)


['churn', 'gender', 'is_senior_citizen', 'partner', 'dependents', 'phone_service', 'paperless_billing']
['tenure', 'multiple_lines', 'internet_service', 'online_security', 'online_backup', 'device_protection', 'tech_support', 'streaming_tv', 'streaming_movies', 'contract', 'payment_method', 'monthly_charges', 'total_charges', 'daily_charges']


Convertendo as colunas tipo object binarios (Yes e Male = 1, No e Female = 0)


In [85]:
dados_binarios = dados_padronizados.copy()
for coluna in lista_col_binaria:
  dados_binarios[coluna] = dados_padronizados[coluna].map({'Yes': 1, 'No': 0, 'Male': 1, 'Female': 0})

Convertendo as colunas tipo object binarios com 3 registros (No =0, Yes = 1, Else = 2)

In [90]:
#lista_col_nao_binaria.remove('daily_charges')
#lista_col_nao_binaria.remove('tenure')
#lista_col_nao_binaria.remove('monthly_charges')
#lista_col_nao_binaria.remove('total_charges')
#lista_col_nao_binaria


for coluna in lista_col_nao_binaria:
  if coluna != 'payment_method' and coluna != 'internet_service' and coluna != 'contract':
    dados_binarios[coluna] = dados_padronizados[coluna].map({'Yes': 1, 'No': 0, 'No internet service': 2, 'No phone service': 2})

# Convertendo as colunas tipo object não binários

contract ('Month-to-month': 0, 'One year': 1, 'Two year': 2)

In [95]:
dados_binarios['contract'] = dados_padronizados['contract'].map({'Month-to-month': 0, 'One year': 1, 'Two year': 2})

internet_service ('Fiber optic': 2, 'DSL': 0, 'No': 1)

In [98]:
dados_binarios['internet_service'] = dados_padronizados['internet_service'].map({'Fiber optic': 2, 'DSL': 1, 'No': 0})

payment_method ('Electronic check': 0, 'Mailed check': 1, 'Bank transfer (automatic)': 2, 'Credit card (automatic)': 3

In [100]:
dados_binarios['payment_method'] = dados_padronizados['payment_method'].map({'Electronic check': 0, 'Mailed check': 1, 'Bank transfer (automatic)': 2, 'Credit card (automatic)': 3})

In [101]:
for coluna in dados_binarios.columns:
  if coluna != 'customer_id' and coluna != 'tenure' and coluna != 'monthly_charges' and coluna != 'total_charges' and coluna != 'daily_charges':
    print(dados_binarios[coluna].value_counts())

churn
0    5163
1    1869
Name: count, dtype: int64
gender
1    3549
0    3483
Name: count, dtype: int64
is_senior_citizen
0    5890
1    1142
Name: count, dtype: int64
partner
0    3639
1    3393
Name: count, dtype: int64
dependents
0    4933
1    2099
Name: count, dtype: int64
phone_service
1    6352
0     680
Name: count, dtype: int64
multiple_lines
0    3385
1    2967
2     680
Name: count, dtype: int64
internet_service
2    3096
1    2416
0    1520
Name: count, dtype: int64
online_security
0    3497
1    2015
2    1520
Name: count, dtype: int64
online_backup
0    3087
1    2425
2    1520
Name: count, dtype: int64
device_protection
0    3094
1    2418
2    1520
Name: count, dtype: int64
tech_support
0    3472
1    2040
2    1520
Name: count, dtype: int64
streaming_tv
0    2809
1    2703
2    1520
Name: count, dtype: int64
streaming_movies
0    2781
1    2731
2    1520
Name: count, dtype: int64
contract
0    3875
2    1685
1    1472
Name: count, dtype: int64
paperless_billing
1    4

In [102]:
dados_binarios.head()

customer_id  churn  gender  is_senior_citizen  partner  dependents  tenure  \
0  0002-ORFBO      0       0                  0        1           1       9   
1  0003-MKNFE      0       1                  0        0           0       9   
2  0004-TLHLJ      1       1                  0        0           0       4   
3  0011-IGKFF      1       1                  1        1           0      13   
4  0013-EXCHZ      1       0                  1        1           0       3   

   phone_service  multiple_lines  internet_service  ...  device_protection  \
0              1               0                 1  ...                  0   
1              1               1                 1  ...                  0   
2              1               0                 2  ...                  1   
3              1               0                 2  ...                  1   
4              1               0                 2  ...                  0   

   tech_support  streaming_tv  streaming_movies  contract  paperless_billing  \
0             1             1                 0         1                  1   
1             0             0                 1         0                  0   
2             0             0                 0         0                  1   
3             0             1                 1         0                  1   
4             1             1                 0         0                  1   

   payment_method  monthly_charges  total_charges  daily_charges  
0               1             65.6         593.30           2.19  
1               1             59.9         542.40           2.00  
2               0             73.9         280.85           2.46  
3               0             98.0        1237.85           3.27  
4               1             83.9         267.40           2.80  

[5 rows x 22 columns]

Salvando dados_binarios

In [103]:
dados_binarios.to_csv('dados_binarios.csv', index=False)

# Carga e analise

## Analise Descritiva

A analise tanto dos dados categoricos quanto dos binarios revela informações pertinentes para as colunas tipo object "binarizadas". Por exemplo é possível descobrir a quantidade de clientes que realizaram churn ao multiplicar o total de clientes pela média de churn: 7032*0.265785 = 1869.
Assim como é possivel observar que há mais clientes do gênero masculino que do feminino. E que mais de 90% dos clientes contratam serviço telefonico da empresa.


In [105]:
url = 'https://raw.githubusercontent.com/mmnasciment/TelecomX_CustomersChurn/refs/heads/main/dados_tratados.csv'
dados_analise = pd.read_csv(url)

In [107]:
dados_analise.describe()

tenure  monthly_charges  total_charges  daily_charges
count  7032.000000      7032.000000    7032.000000    7032.000000
mean     32.421786        64.798208    2283.300441       2.159891
std      24.545260        30.085974    2266.771362       1.002955
min       1.000000        18.250000      18.800000       0.610000
25%       9.000000        35.587500     401.450000       1.187500
50%      29.000000        70.350000    1397.475000       2.340000
75%      55.000000        89.862500    3794.737500       2.992500
max      72.000000       118.750000    8684.800000       3.960000

In [108]:
dados_binarios.describe()

churn       gender  is_senior_citizen      partner   dependents  \
count  7032.000000  7032.000000        7032.000000  7032.000000  7032.000000   
mean      0.265785     0.504693           0.162400     0.482509     0.298493   
std       0.441782     0.500014           0.368844     0.499729     0.457629   
min       0.000000     0.000000           0.000000     0.000000     0.000000   
25%       0.000000     0.000000           0.000000     0.000000     0.000000   
50%       0.000000     1.000000           0.000000     0.000000     0.000000   
75%       1.000000     1.000000           0.000000     1.000000     1.000000   
max       1.000000     1.000000           1.000000     1.000000     1.000000   

            tenure  phone_service  multiple_lines  internet_service  \
count  7032.000000    7032.000000     7032.000000       7032.000000   
mean     32.421786       0.903299        0.615330          1.224118   
std      24.545260       0.295571        0.655867          0.778643   
min       1.000000       0.000000        0.000000          0.000000   
25%       9.000000       1.000000        0.000000          1.000000   
50%      29.000000       1.000000        1.000000          1.000000   
75%      55.000000       1.000000        1.000000          2.000000   
max      72.000000       1.000000        2.000000          2.000000   

       online_security  ...  device_protection  tech_support  streaming_tv  \
count      7032.000000  ...        7032.000000   7032.000000   7032.000000   
mean          0.718857  ...           0.776166      0.722412      0.816695   
std           0.796556  ...           0.778542      0.795571      0.762952   
min           0.000000  ...           0.000000      0.000000      0.000000   
25%           0.000000  ...           0.000000      0.000000      0.000000   
50%           1.000000  ...           1.000000      1.000000      1.000000   
75%           1.000000  ...           1.000000      1.000000      1.000000   
max           2.000000  ...           2.000000      2.000000      2.000000   

       streaming_movies     contract  paperless_billing  payment_method  \
count       7032.000000  7032.000000        7032.000000     7032.000000   
mean           0.820677     0.688567           0.592719        1.315557   
std            0.761287     0.832934           0.491363        1.149523   
min            0.000000     0.000000           0.000000        0.000000   
25%            0.000000     0.000000           0.000000        0.000000   
50%            1.000000     0.000000           1.000000        1.000000   
75%            1.000000     1.000000           1.000000        2.000000   
max            2.000000     2.000000           1.000000        3.000000   

       monthly_charges  total_charges  daily_charges  
count      7032.000000    7032.000000    7032.000000  
mean         64.798208    2283.300441       2.159891  
std          30.085974    2266.771362       1.002955  
min          18.250000      18.800000       0.610000  
25%          35.587500     401.450000       1.187500  
50%          70.350000    1397.475000       2.340000  
75%          89.862500    3794.737500       2.992500  
max         118.750000    8684.800000       3.960000  

[8 rows x 21 columns]

## Distribuição da Evasão

In [178]:
dados_analise.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7032 entries, 0 to 7031
Data columns (total 22 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   customer_id        7032 non-null   object 
 1   churn              7032 non-null   object 
 2   gender             7032 non-null   object 
 3   is_senior_citizen  7032 non-null   object 
 4   partner            7032 non-null   object 
 5   dependents         7032 non-null   object 
 6   tenure             7032 non-null   int64  
 7   phone_service      7032 non-null   object 
 8   multiple_lines     7032 non-null   object 
 9   internet_service   7032 non-null   object 
 10  online_security    7032 non-null   object 
 11  online_backup      7032 non-null   object 
 12  device_protection  7032 non-null   object 
 13  tech_support       7032 non-null   object 
 14  streaming_tv       7032 non-null   object 
 15  streaming_movies   7032 non-null   object 
 16  contract           7032 

Evasão

Churn em % de clientes



In [182]:
import plotly.express as px

fig = px.pie(
    dados_analise,
    names='churn',
    title='Distribuição da Evasão',
    hole=0.6)

fig.update_traces(
    textposition='inside',
    textinfo='percent+label+value',
    hovertemplate='<b>%{label}</b><br>Clientes: %{value}<br>Porcentagem: %{percent}<extra></extra>' # Texto do hover
)

fig.show()

### Variaveis Numéricas

Dos churn positivos, quanto tempo ficaram na empresa?

In [185]:
tenure_igual_12 = churn_sim[churn_sim['tenure'] <= 12]

tenure_igual_12['tenure_label'] = tenure_igual_12['tenure'].apply(
    lambda x: f'{x} Mês' if x == 1 else f'{x} Meses')


fig = px.pie(
    tenure_igual_12,
    names='tenure_label',
    title='Churn positivo por tenure (clientes que ficaram até 1 ano com a empresa)',
    hole=0.6,
    category_orders={'tenure_label': [f'{i} Mês' if i == 1 else f'{i} Meses' for i in range(1, 13)]}
)

fig.update_traces(
    textposition='inside',
    textinfo='percent+label+value',
    hovertemplate='<b>%{label}</b><br>Clientes: %{value}<br>Porcentagem: %{percent}<extra></extra>' # Texto do hover
)

fig.show()

/tmp/ipython-input-185-2329530349.py:3: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



Dos Churn positivo, quanto o cliente gastou?

Analisando os dados estatisticos de monthly_charges para montar um histograma que faça sentido para o contexto.

In [201]:
dados_analise['monthly_charges'].mode()

0    20.05
Name: monthly_charges, dtype: float64

In [199]:
dados_analise.describe()

tenure  monthly_charges  total_charges  daily_charges
count  7032.000000      7032.000000    7032.000000    7032.000000
mean     32.421786        64.798208    2283.300441       2.159891
std      24.545260        30.085974    2266.771362       1.002955
min       1.000000        18.250000      18.800000       0.610000
25%       9.000000        35.587500     401.450000       1.187500
50%      29.000000        70.350000    1397.475000       2.340000
75%      55.000000        89.862500    3794.737500       2.992500
max      72.000000       118.750000    8684.800000       3.960000

In [213]:
condicao2 = dados_analise['churn'] == 'Yes'

churn_mensal = dados_analise.loc[condicao2]

fig = px.histogram(
    churn_mensal,
    x='monthly_charges',
    title='Distribuição de gasto dos Clientes com Churn positivo',
    labels={'tenure': 'Tempo de Contrato (em meses)', 'count': 'Número de Clientes'},
    range_x=[0, 120],
    marginal='rug',
    nbins=8
)

fig.update_layout(
    yaxis_title='Frequência (Nº de Clientes)',
    xaxis_title='Total Gasto (USD)'
)

fig.show()

### Variaveis Categoricas

Dos Churn positivos, qual o tipo de contrato?

In [184]:
churn_sim = dados_analise[dados_analise['churn'] == 'Yes']

fig = px.pie(
    churn_sim,
    names='contract',
    title='Churn por tipo de contrato',
    hole=0.6
)
fig.update_traces(
    textposition='inside',
    textinfo='percent+label+value',
    hovertemplate='<b>%{label}</b><br>Clientes: %{value}<br>Porcentagem: %{percent}<extra></extra>' # Texto do hover
)

fig.show()

Dos Churn positivos, qual era o metodo de pagamento?

In [123]:
fig = px.pie(
    churn_sim,
    names='payment_method',
    title='Churn por tipo de pagamento',
    hole=0.6
)


fig.show()

Dos Churn positivo, qual serviço utilizaram?

In [173]:
tenure_menor_que_12_servicos = tenure_igual_12[['phone_service', 'multiple_lines', 'internet_service', 'online_security', 'online_backup', 'device_protection', 'tech_support', 'streaming_tv', 'streaming_movies']]


In [174]:
fig = px.pie(
    tenure_menor_que_12_servicos,
    names='phone_service',
    title='Churn positivo com até 1 ano de empresa que contrataram serviço telefonico',
    hole=0.6
)

fig.update_traces(
    textposition='inside',
    textinfo='percent+label+value',
    #hovertemplate='<b>%{label}</b><br>Clientes: %{value}<br>Porcentagem: %{percent}<extra></extra>' # Texto do hover
)


fig.show()